src/WebJobs.Extensions.DurableTask/ContextImplementations/DurableOrchestrationContext.cs

```diff
@@ -308,22 +310,45 @@ async Task<T> IDurableOrchestrationContext.CreateTimer<T>(DateTime fireAt, T sta
         {
             this.ThrowIfInvalidAccess();
 
-            if (!this.durabilityProvider.ValidateDelayTime(fireAt.Subtract(this.InnerContext.CurrentUtcDateTime), out string errorMessage))
+            DateTime intervalFireAt = fireAt;
+
+            if (fireAt.Subtract(this.InnerContext.CurrentUtcDateTime) > this.durabilityProvider.MaximumDelayTime)
             {
-                throw new ArgumentException(errorMessage, nameof(fireAt));
+                this.IsLongRunningTimer = true;
+                intervalFireAt = this.InnerContext.CurrentUtcDateTime.Add(this.durabilityProvider.LongRunningTimerIntervalLength);
             }
 
-            this.IncrementActionsOrThrowException();
-            Task<T> timerTask = this.InnerContext.CreateTimer(fireAt, state, cancelToken);
-
             this.Config.TraceHelper.FunctionListening(
                 this.Config.Options.HubName,
                 this.OrchestrationName,
                 this.InstanceId,
                 reason: $"CreateTimer:{fireAt:o}",
                 isReplay: this.InnerContext.IsReplaying);
 
-            T result = await timerTask;
+            T result = default;
+
+            while (this.InnerContext.CurrentUtcDateTime < fireAt && !cancelToken.IsCancellationRequested)
+            {
+                this.IncrementActionsOrThrowException();
+                Task<T> timerTask = this.InnerContext.CreateTimer(intervalFireAt, state, cancelToken);
+
+                result = await timerTask;
+
+                TimeSpan remainingTime = fireAt.Subtract(this.InnerContext.CurrentUtcDateTime);
+                if (remainingTime <= TimeSpan.Zero)
+                {
+                    break;
+                }
+                else if (remainingTime < this.durabilityProvider.LongRunningTimerIntervalLength)
+                {
+                    TimeSpan timerStartsIn = fireAt.Subtract(this.InnerContext.CurrentUtcDateTime);
+                    intervalFireAt = this.InnerContext.CurrentUtcDateTime.Add(timerStartsIn);
```

Comment:
Are we doing redundant math here? I would have expected the following:
```suggestion
                    intervalFireAt = this.InnerContext.CurrentUtcDateTime.Add(remainingTime);
```

...and removing line 344.